### DISCUSSION TITLE
Convert a url with lazyload image part

### DISCUSSION THREAD

**Discussion by RandalTeng** [gotenberg/gotenberg#742] [Q&A]
I have a issue about lazyload image.
My boss want convert a url to pdf when the sever receiver a request with a third-part url, the url has a lazyload image part.
If the url content more than scoll hight, after the pdf is converted, there is lot of blank image which is pre-load image.
How could I make Gotenberg load the truely image?

**Comment by gulien** [comment]
Hello @RandalTeng,

A good option would be to inject some script that scroll to the end of the page (to force the lazy-loading of images) and once its is done, set a global variable that can be catch by `waitForExpression` (https://gotenberg.dev/docs/routes#wait-before-rendering).

With a third-part URL, it may be complicated though.

**Reply by RandalTeng** [reply]
Thanks for the reply first, how ever...
Emmm, there is a problem more...
How could I inject the scroll script to the page inside of Gotenberg's Chromium instance?
By the way, our Gotenberg instance is run in container base on K8s platform.

**Reply by RandalTeng** [reply]
Is the `waitForExpression` attribute support a full script expression?
Like set the script content to this attribute, and return a true flag as final signal?

**Reply by gulien** [reply]
Huh, it could work lol please let me know 😄

**Reply by RandalTeng** [reply]
Okay, I can't get it.
I had tried set expression to `waitForExpression`, but it seem can't handle wait a promise, when scoll the content, program need a little time to wait dispatch events.
Is there any other option to trigger lazy-loading?

**Reply by gulien** [reply]
Yeah, I was expecting this...
Currently not, maybe try to disable JavaScript with  `--chromium-disable-javascript`?

**Reply by RandalTeng** [reply]
I'm not sure about is the image load time make the image can't be load, or the js script not run as except.
I have look for `waitForExpression` implement code, when this option(function?) run completed, the chromium convert content to pdf immediately, I have tried set `waitForDelay` in it too, but the source code show that `waitForDelay` is running before `waitForExpression`, it can't be work.

I will download source code, and change the process flow, make a test for it.
Reply this for a record...

**Reply by RandalTeng** [reply]
emm, @gulien here need some help..
I had change the `waitDelay` after `waitForExpression` to be execute, and the lazy-load image is loaded when the convert route execute first time, after this, all image want be loaded.
I can't figure out where is wrong, my change in here: 
[https://github.com/RandalTeng/gotenberg/blob/feature/rt-wait-attr-flow/pkg/modules/chromium/browser.go#L267](https://github.com/RandalTeng/gotenberg/blob/feature/rt-wait-attr-flow/pkg/modules/chromium/browser.go#L267)
Ps: and after convert, I had add a ClosePage call ( maybe there is a memory limit I think, in the end, it not work, so ignore it. )

**Reply by gulien** [reply]
Hello @RandalTeng,

I think your best option for now is to use a library like [Puppeteer](https://pptr.dev/) to programmatically handle your use case. Gotenberg may be too complicated to modify to your flavor if you're not familiar with Go or the Chrome DevTools Protocol.

**Reply by RandalTeng** [reply]
Sorry for disturbing.
Actually, I want to know what happened when the convert executed.
I had look for logs that the truly image has been requested and responsed, but when convert output, the pdf just display the loading image, I can't get it.

Sorry for disturbing again, I will find it out alone, thanks for your reply.
